nova/objects/instance.py

```diff
@@ -1634,12 +1635,18 @@ def _get_instance_types(context, wanted_itypes):
 
     @staticmethod
     def _sum_counts(instances, instance_types):
-        counts = {'instances': 0, 'cores': 0, 'ram': 0}
+        return InstanceList._sum_counts_by(instances, instance_types)[()]
+
+    @staticmethod
+    def _sum_counts_by(instances, instance_types):
+        counts_by_fields = collections.defaultdict(
+            lambda: {'instances': 0, 'cores': 0, 'ram': 0})
 
         if not instances:
-            return counts
+            return counts_by_fields
 
-        for type_id, instance_count, cores, ram in instances:
+        for type_id, instance_count, cores, ram, *rest in instances:
+            counts = counts_by_fields[tuple(rest)]
```

Comment:
Maybe "group_by", "group_by_key", "group_keys" or "fields" would be a better name for `rest`?

Reply:
I'll go with `group_keys`. I wanted `fields` but that was somehow not possible (no idea why).